### PR TITLE
add prefetch module && mv Mithril to prefetch

### PR DESCRIPTION
**Pull request by zztaki**
Based on [the issue](https://github.com/1a1a11a/libCacheSim/issues/16).

### CODE REVIEW

**Comment by 1a1a11a** [comment]
This is great work! Thank you! 
One question is whether we should place Mithril under eviction. I was planning to have a prefetching module (we already have eviction and admission) so that we can use different combinations of prefetching and eviction algorithms.  Do you want to take a crack?

**Comment by zztaki** [comment]
> This is great work! Thank you!
> One question is whether we should place Mithril under eviction. I was planning to have a prefetching module (we already have eviction and admission) so that we can use different combinations of prefetching and eviction algorithms. Do you want to take a crack?

Yeah, taking the long view, it will be better! 

In my understanding, if we plan to make a separate prefetch module, we need to save the prefetcher as a parameter in cache_t and init it in cache_init like the admission module. In the execution of the cache algorithm, query the prefetch_table in cache_get_base for prefetching, and record or mining in some other basic functions, such as cache_evict_base. And then generate prefetch_table periodically.

I don't know if I get your suggestion. My PC is not around at the moment, so these words may be a little confusing. hh

**Comment by zztaki** [comment]
> You got it right! That is almost the same as my plan: we can add a `prefetch(cache_t *cache, requests_t *req)` function to cache_t and set the default value to `NULL`, if it is not NULL, we prefetch. :)

Oh, I know! I am glad to add a prefetching module and than add the mithril algo!

**Comment by zztaki** [comment]
Hi, I add a prefetching module and move the Mithril algo to it!

The only thing that is confusing is this code paragraph.
```c
bool cache_get_base(cache_t *cache, const request_t *req) {
  cache->n_req += 1;

  VERBOSE("******* %s req %ld, obj %ld, obj_size %ld, cache size %ld/%ld\n",
          cache->cache_name, cache->n_req, req->obj_id, req->obj_size,
          cache->get_occupied_byte(cache), cache->cache_size);

  cache_obj_t *obj = cache->find(cache, req, true);
  bool hit = (obj != NULL);

  if (hit) {
    VVERBOSE("req %ld, obj %ld --- cache hit\n", cache->n_req, req->obj_id);
  }

  if (cache->can_insert(cache, req) == false) {
    VVERBOSE("req %ld, obj %ld --- cache miss cannot insert\n", cache->n_req,
             req->obj_id);
    return false;
  }

  if (cache->prefetcher) {
    // this interface should do the following

    // if !hit:
    //     evict until have enough space to insert the requested obj
    //     insert the requested obj
    // prefetch
    // evict until cache is not full
    cache->prefetcher->insert_prefetch(cache, req, !hit);
  } else if (!hit) {
    while (cache->get_occupied_byte(cache) + req->obj_size +
               cache->obj_md_size >
           cache->cache_size) {
      cache->evict(cache, req);
    }
    cache->insert(cache, req);
  }

  return hit;
}
```

File: test/common.h [hunk 1]
```diff
@@ -232,6 +234,10 @@ static cache_t *create_test_cache(const char *alg_name,
     cache = S3FIFO_init(cc_params, "move-to-main-threshold=2");
   } else if (strcasecmp(alg_name, "Sieve") == 0) {
     cache = Sieve_init(cc_params, NULL);
+  } else if (strcasecmp(alg_name, "Mithril") == 0) {
```
Comment (1a1a11a) [inline test/common.h:237]: Does this need to be updated to use `-p` parameter?

**Comment by 1a1a11a** [comment]
> Hi, I add a prefetching module and move the Mithril algo to it!
> 
> The only thing that is confusing is this code paragraph.
> 
> ```c
> bool cache_get_base(cache_t *cache, const request_t *req) {
>   cache->n_req += 1;
> 
>   VERBOSE("******* %s req %ld, obj %ld, obj_size %ld, cache size %ld/%ld\n",
>           cache->cache_name, cache->n_req, req->obj_id, req->obj_size,
>           cache->get_occupied_byte(cache), cache->cache_size);
> 
>   cache_obj_t *obj = cache->find(cache, req, true);
>   bool hit = (obj != NULL);
> 
>   if (hit) {
>     VVERBOSE("req %ld, obj %ld --- cache hit\n", cache->n_req, req->obj_id);
>   }
> 
>   if (cache->can_insert(cache, req) == false) {
>     VVERBOSE("req %ld, obj %ld --- cache miss cannot insert\n", cache->n_req,
>              req->obj_id);
>     return false;
>   }
> 
>   if (cache->prefetcher) {
>     // this interface should do the following
> 
>     // if !hit:
>     //     evict until have enough space to insert the requested obj
>     //     insert the requested obj
>     // prefetch
>     // evict until cache is not full
>     cache->prefetcher->insert_prefetch(cache, req, !hit);
>   } else if (!hit) {
>     while (cache->get_occupied_byte(cache) + req->obj_size +
>                cache->obj_md_size >
>            cache->cache_size) {
>       cache->evict(cache, req);
>     }
>     cache->insert(cache, req);
>   }
> 
>   return hit;
> }
> ```

"insert the object and then do prefetching and evicting" should be fine now. But merging insert and prefetch is a bit weird. Do you think we can separate them into different functions? We can do a prefetcher update in `insert_base` as well.

**Comment by 1a1a11a** [comment]
BTW, this is really great work! I am happy to merge it if you do not want to change anymore. But if we could have a well-defined interface for prefetch, more users will benefit. And my suggestion sometimes may not be reasonable, feel free to share your thoughts. :)

File: test/common.h (hunk 1 above)
Comment (zztaki) [reply test/common.h:237]: I intended to add a Mithril test to this place, and add a `testPrefetchAlgo.c` file to test all prefetch algorithms, but due to time reasons, I did not complete the test work.

Comment (1a1a11a) [reply test/common.h:237]: Sounds good! Will you have time to finish (no rush)? 
BTW, thank you for fixing the leak in admission module :)

Comment (zztaki) [reply test/common.h:237]: Yes, I think I can try it! But need a little time.

Comment (1a1a11a) [reply test/common.h:237]: Great! There is no rush! Let me know if you need any help from me.

**Comment by zztaki** [comment]
> BTW, this is really great work! I am happy to merge it if you do not want to change anymore. But if we could have a well-defined interface for prefetch, more users will benefit. And my suggestion sometimes may not be reasonable, feel free to share your thoughts. :)

Yeah, separating is great.

I think we can add two extra interfaces in `prefetch_t`, such as `update_at_insert` and `update_at_evict`. Eh, for `Mithrim`, it maybe need to update_at_evict during `cache_evict_base` in my understanding.

**Comment by zztaki** [comment]
Hi, I add some prefetcher interfaces, such as `handle_find` and `handle_evict`. 

In the future, if some prefetch algorithms need `handle_insert`, it can be added in proper place!

**Comment by zztaki** [comment]
It is worth noting that I call `prefetcher->handle_find` interface before `cache->find` in `cache_get_base` instead of `cache_find_base`, because the prefetch algorithm needs to call `cache->find` in `handle_find` to check, if not add a new parameter in `cache->find`, it will fall into an infinite loop.

It would be great if you have better advice!

**Comment by zztaki** [comment]
Is there any problem? :)

**Comment by 1a1a11a** [comment]
Sorry I was traveling, will take a look as soon as possible

**Comment by 1a1a11a** [comment]
Thank you on the great work! :)

**Comment by zztaki** [comment]
Wonderful!

**Pull request merged**